From: Jiri Slaby <jslaby@suse.cz>
Date: Wed, 7 Nov 2012 15:34:30 +0100
Subject: x86: revert HLT removal
Patch-mainline: never, kabi fix

In 3.0.51, commit 05e02741ed7 (x86: Remove the ancient and deprecated
disable_hlt() and enable_hlt() facility), upstream commit
f6365201d8a21f HLT functions were removed. but these were part of our
kABI on x86_32. Add them back.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/include/asm/system.h |  7 +++++++
 arch/x86/kernel/process.c     | 24 ++++++++++++++++++++++++
 2 files changed, 31 insertions(+)

diff --git a/arch/x86/include/asm/system.h b/arch/x86/include/asm/system.h
index f0d89d9..c2ff2a1 100644
--- a/arch/x86/include/asm/system.h
+++ b/arch/x86/include/asm/system.h
@@ -93,6 +93,10 @@ do {									\
 			"memory");					\
 } while (0)
 
+/*
+ * disable hlt during certain critical i/o operations
+ */
+#define HAVE_DISABLE_HLT
 #else
 
 /* frame pointer must be last for get_wchan */
@@ -388,6 +392,9 @@ static inline void clflush(volatile void *__p)
 
 #define nop() asm volatile ("nop")
 
+void disable_hlt(void);
+void enable_hlt(void);
+
 void cpu_idle_wait(void);
 
 extern unsigned long arch_align_stack(unsigned long sp);
diff --git a/arch/x86/kernel/process.c b/arch/x86/kernel/process.c
index 4272502..e1ba8cb 100644
--- a/arch/x86/kernel/process.c
+++ b/arch/x86/kernel/process.c
@@ -341,10 +341,34 @@ void (*pm_idle)(void);
 EXPORT_SYMBOL(pm_idle);
 #endif
 
+#ifdef CONFIG_X86_32
+/*
+ * This halt magic was a workaround for ancient floppy DMA
+ * wreckage. It should be safe to remove.
+ */
+static int hlt_counter;
+void disable_hlt(void)
+{
+	hlt_counter++;
+}
+EXPORT_SYMBOL(disable_hlt);
+
+void enable_hlt(void)
+{
+	hlt_counter--;
+}
+EXPORT_SYMBOL(enable_hlt);
+
+static inline int hlt_use_halt(void)
+{
+	return (!hlt_counter && boot_cpu_data.hlt_works_ok);
+}
+#else
 static inline int hlt_use_halt(void)
 {
 	return 1;
 }
+#endif
 
 /*
  * We use this if we don't have any better
-- 
1.8.0

